### PR TITLE
MM-24009: remove redux-offline

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
See <PR> for details, with this PR containing the necessary changes to support removing redux-offline as a library.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-24009

### CODE REVIEW

**Comment by lieut-data** [comment]
I'm not yet sure why this is failing in CI, as the build passes locally. Will investigate the typing issues.

File: store/index.js [hunk 1]
```diff
@@ -195,21 +194,43 @@ export default function configureStore(initialState) {
...
                 return state[key];
             },
             _stateSetter: (state, key, value) => {
+                // eslint-disable-next-line no-process-env
+                if (process.env.NODE_ENV === 'production') {
+                    if (key.indexOf('storage:') === 0) {
+                        state.storage = state.storage || {storage: {}};
+                        state.storage.storage[key.substr(8)] = value;
+                    } else {
+                        state[key] = value;
```
Comment (lieut-data) [inline store/index.js:208]: Note that previously `state[key] = value` occurred /in addition/ to the redirected overwrite. I tested this and couldn't find any reason to do this, so removed the spurious write.

File: store/index.js [hunk 2]
```diff
@@ -195,21 +194,43 @@ export default function configureStore(initialState) {
             },
             _stateGetter: (state, key) => {
                 if (key.indexOf('storage:') === 0) {
-                    state.storage = state.storage || {storage: {}};
-                    return state.storage.storage[key.substr(8)];
+                    return state.storage?.storage?.[key.substr(8)];
```
Comment (lieut-data) [inline store/index.js:197]: Avoid mutating the store at all here.

Comment (hmhealey) [reply store/index.js:197]: I'm surprised that didn't cause us any problems before since that should've caused it to error out

**Review by hmhealey — APPROVED** [review]

**Review by devinbinnie — APPROVED** [review]

**Comment by devinbinnie** [comment]
@lieut-data Does this need QA? Not sure there's much for QA to review here, perhaps on the webapp PR?

**Comment by lieut-data** [comment]
Good call, @devinbinnie -- nothing for QA review on /this/ PR. Let's focus our efforts on the webapp portion.

**Comment by lieut-data** [comment]
(Just realized this /is/ the webapp PR, not the redux one :p)

**Comment by jgilliam17** [comment]
@lieut-data You mentioned in the linked ticket "posting, deleting posts, editing preferences, leaving a channel and retry posting should it fail" Are there any other areas I should concentrate testing on? Let me know. Thanks.

**Comment by lieut-data** [comment]
@jgilliam17, I think that's all the areas that are "touched" by this change, having previously relied on redux-offline for some extra behaviour. Thanks for double checking! :)

**Comment by jgilliam17** [comment]
@lieut-data seeing few errors in console
Screen Shot 2021-03-11 at 9 05 29 AM

**Comment by lieut-data** [comment]
@jgilliam17, yikes that server seems to be spewing gateway errors. Do you know the context in which that happened? It /seems/ like a transient cloud issue, maybe because the server went down overnight for some maintenance?

**Comment by jgilliam17** [comment]
I believe I was mentioning a user (not currently a member of the channel) and then adding them to channel when a whole page of errors loaded

**Comment by lieut-data** [comment]
Got it, @jgilliam17 -- given the myriad of endpoints failing there, I suspect it's something cloud specific. Can you raise it with the cloud team just so we can be sure to trace down the "hiccup" here? Also, I suspect this is not reproducible?

**Comment by jgilliam17** [comment]
Not reproducible. I cleared the console and so far see only one line for incident plugin 
Screen Shot 2021-03-11 at 9 18 03 AM
Also, this is probably a Chrome issue, when throttling network and going `Offline` I am still receiving messages, which is weird, but can't send (as expected). 
Screen Shot 2021-03-11 at 9 19 09 AM 1

**Comment by lieut-data** [comment]
Sounds good -- the incident collaboration is expected, in that 404s are being used semantically vs. indicative of an error.

Yes, chrome is unusual in that their "offline mode" doesn't stop active websockets, so incoming messages still get processed. It's caught me off guard several times :)

**Comment by jgilliam17** [comment]
@lieut-data I've reviewed [E2E resutls](https://mm-cypress-report.s3.amazonaws.com/3743-webapp-pr-7648-mattermost-enterprise-edition:9131b3e/mochawesome.html), compared with latest master prod test run; most failures are timed out tests and not repro-ing when manually tested on the test server (or have a fix pending)
There is an error and perpetually loading screen when leaving an archived channel; reproduces manually on the test server. 
Screen Shot 2021-03-12 at 11 25 44 AM

**Comment by lieut-data** [comment]
Thanks for your careful eye, @jgilliam17! Can you help me with reproducing the above archived channel oddity? I've tested leaving archived channels in a variety of ways both locally and on the test server for this PR without encountering the specific issue seen above.

Perhaps it requires a particular setup before getting stuck in this state?

**Comment by jgilliam17** [comment]
Odd, it's intermittent for me now.
On the PR server - try searching for an archived channel you are not a member off (as sysadmin, `mmm` channel), click on the name to join and then use channel menu option to Leave Channel

**Comment by lieut-data** [comment]
This is a tremendous find, @jgilliam17! So when things seem non-deterministic like this, I often find that throttling the network tends to surface them more easily (i.e. via Chrome/Firefox devtools). In this case, simulating a 2G connection allowed me to recreate the problem quite easily, /and/ verify that it's not occurring on `master`.

Looking at the code again, I observed a subtle semantic difference to my changes: the parent action now waits for the Client4 API call to finish (or fail) before returning to the caller. Previously, by virtue of being embedded within redux-offline magic, it would occur asynchronously.

It turns out some of the webapp code is sensitive to this (probably incorrectly), and ends up racing with the code to determine the next steps, incorrectly sending the user back to the channel they just left since the Client4 API call leaving the channel probably results in a websocket event or some such and a later event.

To avoid this, I've restored the original behaviour by wrapping everything in an immediately-evaluated-function-expression (just to keep the `async` capabilities) but on which the parent action does not wait. I could not reproduce the behaviour after making this change.

I've asked for a second dev review, and will loop you back in once that passes muster. Thanks again :)

**Comment by lieut-data** [comment]
(For clarity, the changes to review are in https://github.com/mattermost/mattermost-redux/pull/1396).

**Review by devinbinnie — APPROVED** [review]

**Review by hmhealey — APPROVED** [review]

**Review by jgilliam17 — APPROVED** [review]
Thanks @lieut-data - Let me know if you think we need to run E2Es again
Quick smoke test and it looks good to merge ( tested with no throttling and slow 3G)
- Verified posting, deleting posts, editing preferences, leaving public, private or archived channels and `retry` posting after connection loss - no issues.

**Pull request merged**